File: cpu_usage_indicator/go/led.go
package main

import "fmt"
import "os"


type led_t struct {
	devpath string
	brightness uint8
}


func led_init(led *led_t, devpath string) {
	led.devpath = devpath

	led_set_brightness(led, 0)
}


func led_fini(led *led_t) {
	led_set_brightness(led, 0)
}


func led_set_brightness(led *led_t, brightness uint8) {
	file, err := os.OpenFile(led.devpath, os.O_WRONLY, 0664)
	if err != nil {
		panic(err)
	}

	fmt.Fprintf(file, "%d\n", int(brightness))

	err = file.Close()
	if err != nil {
		panic(err)
	}

	led.brightness = brightness
}

File: cpu_usage_indicator/c/led.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <string.h>

#include "led.h"

void led_init(led_t* led, const char* devpath, unsigned char brightness)
{
	strncpy(led->devpath, devpath, sizeof(led->devpath));
	led->devpath[sizeof(led->devpath) - 1] = '\0';

	led_set_brightness(led, brightness);
}


void led_fini(led_t* led)
{
    led_set_brightness(led, 0);
}


void led_set_brightness(led_t* led, unsigned char brightness)
{
	FILE* f = fopen(led->devpath, "w");
	assert(f != NULL);
	fprintf(f, "%d\n", brightness);
	fclose(f);

	led->brightness = brightness;
}

